samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemRemoveListener.java

```diff
@@ -0,0 +1,145 @@
+  //private String scalename;  // used for multiple choice Survey
+
+  /**
+   * Standard process action method.
+   * @param ae ActionEvent
+   * @throws AbortProcessingException
+   */
+  public void processAction(ActionEvent ae) throws AbortProcessingException
+  {		
+	  ItemAuthorBean item = (ItemAuthorBean) ContextUtil.lookupBean("itemauthor");
+	  ItemService delegate = new ItemService();
+	  if(item.getItemToDelete() == null) return;
+	  Long deleteId= item.getItemToDelete().getItemId();
+	  ItemFacade itemf = delegate.getItem(String.valueOf(deleteId));
+	  // save the currSection before itemf.setSection(null), used to reorder question sequences
+	  SectionFacade  currSection = (SectionFacade) itemf.getSection();
+	  Integer  currSeq = itemf.getSequence();
+
+	  QuestionPoolService qpdelegate = new QuestionPoolService();
+	  if (qpdelegate.getPoolIdsByItem(deleteId.toString()) ==  null || qpdelegate.getPoolIdsByItem(deleteId.toString()).isEmpty()){
+		  // if no reference to this item at all, ie, this item is created in assessment but not assigned to any pool
+
+		  AuthorizationBean authzBean = (AuthorizationBean) ContextUtil.lookupBean("authorization");
+		  AssessmentService assessdelegate = new AssessmentService();
+		  AssessmentFacade af = assessdelegate.getBasicInfoOfAnAssessmentFromSectionId(currSection.getSectionId());
+		  if (!authzBean.isUserAllowedToEditAssessment(af.getAssessmentBaseId().toString(), af.getCreatedBy(), false)) {
+		      throw new IllegalArgumentException("User does not have permission to delete item in assessment: " + af.getAssessmentBaseId());
+		  }
+
+		  delegate.deleteItem(deleteId, AgentFacade.getAgentString());
+	  }
+	  else {
+		  if (currSection == null) {
+			  // if this item is created from question pool
+			  QuestionPoolBean  qpoolbean= (QuestionPoolBean) ContextUtil.lookupBean("questionpool");
+	          ItemFacade itemfacade= delegate.getItem(String.valueOf(deleteId));
+	          ArrayList items = new ArrayList();
+	          items.add(itemfacade);
+	          qpoolbean.setItemsToDelete(items);
+			  qpoolbean.removeQuestionsFromPool();
+			  return;
+		  }
+		  else {
+			  // 
+			  // if some pools still reference to this item, ie, this item is 
+			  // created in assessment but also assigned a a pool
+			  // then just set section = null
+			  itemf.setSection(null);
+			  delegate.saveItem(itemf);
+		  }
+	  }
+	//An item has been deleted
+	EventTrackingService.post(EventTrackingService.newEvent("sam.assessment.item.delete", "/sam/" +AgentFacade.getCurrentSiteId() + "/removed itemId=" + deleteId, true));
+
+	AssessmentService assessdelegate = new AssessmentService();
+      // reorder item numbers
+
+	  SectionFacade sectfacade = assessdelegate.getSection(currSection.getSectionId().toString());
+	  Set itemset = sectfacade.getItemFacadeSet();
+	  // should be size-1 now.
+	  Iterator iter = itemset.iterator();
```

Comment:
Iterator not necessary here, you can use a `for()` loop over the items in the `Set<>`